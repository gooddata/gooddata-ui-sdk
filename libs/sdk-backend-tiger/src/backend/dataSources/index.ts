// (C) 2021-2024 GoodData Corporation

import { IDataSourcesService } from "@gooddata/sdk-backend-spi";
import { IDataSourceIdentifierDescriptor } from "@gooddata/sdk-model";
import { TigerAuthenticatedCallGuard } from "../../types/index.js";
import { ActionsUtilities, JsonApiDataSourceIdentifierOutWithLinks } from "@gooddata/api-client-tiger";

function convertdataSourceIdentifier(
    dataSource: JsonApiDataSourceIdentifierOutWithLinks,
): IDataSourceIdentifierDescriptor {
    return {
        id: dataSource.id,
        name: dataSource.attributes.name,
        schema: dataSource.attributes.schema,
        type: dataSource.attributes.type,
    };
}

export class TigerDataSourcesService implements IDataSourcesService {
    constructor(private readonly authCall: TigerAuthenticatedCallGuard) {}

    public async getDataSourceIdentifiers(): Promise<IDataSourceIdentifierDescriptor[]> {
        return this.authCall(async (client) => {
            return ActionsUtilities.loadAllPages(({ page, size }) =>
                client.entities
                    .getAllEntitiesDataSourceIdentifiers({ page, size })
                    .then((response) => response.data.data.map(convertdataSourceIdentifier)),
            );
        });
    }
}
