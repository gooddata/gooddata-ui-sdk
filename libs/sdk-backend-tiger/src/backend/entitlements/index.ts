// (C) 2021-2024 GoodData Corporation

import { IEntitlements } from "@gooddata/sdk-backend-spi";
import { IEntitlementDescriptor } from "@gooddata/sdk-model";
import { TigerAuthenticatedCallGuard } from "../../types/index.js";

export class TigerEntitlements implements IEntitlements {
    constructor(private readonly authCall: TigerAuthenticatedCallGuard) {}

    public async resolveEntitlements(): Promise<IEntitlementDescriptor[]> {
        const profile = await this.authCall((client) => client.profile.getCurrent());
        return (
            profile.entitlements ??
            (await this.authCall((client) => client.actions.resolveAllEntitlements())).data
        );
    }
}
