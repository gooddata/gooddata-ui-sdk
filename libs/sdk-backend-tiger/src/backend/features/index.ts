// (C) 2020-2024 GoodData Corporation
import {
    IStaticFeatures,
    IUserProfile,
    ILiveFeatures,
    FeatureContext,
    JsonApiWorkspacePatchAttributes,
} from "@gooddata/api-client-tiger";
import { LRUCache } from "lru-cache";
import { TigerAuthenticatedCallGuard } from "../../types/index.js";
import { ITigerFeatureFlags, DefaultFeatureFlags } from "../uiFeatures.js";

import { getFeatureHubFeatures } from "./hub.js";
import { getStaticFeatures } from "./static.js";

const getKeyFromContext = (wsContext?: Partial<FeatureContext>): string => {
    return `${wsContext?.organizationId}-${wsContext?.earlyAccess}`;
};
const responseMap: LRUCache<string, Promise<ITigerFeatureFlags>> = new LRUCache<
    string,
    Promise<ITigerFeatureFlags>
>({
    max: 10,
});

export class TigerFeaturesService {
    constructor(private readonly authCall: TigerAuthenticatedCallGuard) {}

    public async getFeatures(
        profile?: IUserProfile,
        wsContext?: Partial<FeatureContext>,
    ): Promise<ITigerFeatureFlags> {
        const cachedResponse = responseMap.get(getKeyFromContext(wsContext));
        if (cachedResponse) {
            return cachedResponse;
        }
        const response = this.authCall(async (client) => {
            const prof = profile || (await client.profile.getCurrent());
            const results = await loadFeatures(prof, wsContext);

            return {
                ...DefaultFeatureFlags,
                ...results,
            };
        });
        responseMap.set(getKeyFromContext(wsContext), response);
        return response;
    }
}

async function loadFeatures(
    profile: IUserProfile,
    wsContext: Partial<FeatureContext> = {},
): Promise<Partial<ITigerFeatureFlags>> {
    const features = profile.features || {};

    if (featuresAreLive(features)) {
        return await getFeatureHubFeatures(features.live, wsContext);
    }

    if (featuresAreStatic(features)) {
        return await getStaticFeatures(features.static);
    }
    return Promise.resolve({});
}

function featuresAreLive(item: any): item is ILiveFeatures {
    return Boolean(item?.live);
}

function featuresAreStatic(item: any): item is IStaticFeatures {
    return Boolean(item?.static);
}

export function pickContext(
    attributes: JsonApiWorkspacePatchAttributes | undefined,
    organizationId: string | undefined,
): Partial<FeatureContext> {
    const context: Partial<FeatureContext> = {};

    if (attributes?.earlyAccess !== undefined) {
        context.earlyAccess = attributes.earlyAccess;
    }

    if (organizationId !== undefined) {
        context.organizationId = organizationId;
    }
    return context;
}
