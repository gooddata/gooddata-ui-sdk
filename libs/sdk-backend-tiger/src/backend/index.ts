// (C) 2019-2024 GoodData Corporation
import { AxiosInstance, AxiosResponse } from "axios";
import { invariant } from "ts-invariant";
import {
    IAnalyticalBackendConfig,
    IBackendCapabilities,
    IAnalyticalBackend,
    IAnalyticalWorkspace,
    IAuthenticationProvider,
    IAuthenticatedPrincipal,
    IWorkspacesQueryFactory,
    IUserService,
    ErrorConverter,
    NotAuthenticated,
    IAuthenticationContext,
    isNotAuthenticated,
    IOrganization,
    IOrganizations,
    IEntitlements,
    isContractExpired,
    IDataSourcesService,
} from "@gooddata/sdk-backend-spi";
import { newAxios, tigerClientFactory, ITigerClient } from "@gooddata/api-client-tiger";
import isEmpty from "lodash/isEmpty.js";
import isError from "lodash/isError.js";
import isString from "lodash/isString.js";
import inRange from "lodash/inRange.js";
import identity from "lodash/identity.js";
import omit from "lodash/omit.js";

import { convertApiError } from "../utils/errorHandling.js";

import { TigerWorkspace } from "./workspace/index.js";
import { TigerWorkspaceQueryFactory } from "./workspaces/index.js";
import { TigerUserService } from "./user/index.js";
import {
    AuthProviderCallGuard,
    AuthenticatedAsyncCall,
    IAuthenticatedAsyncCallContext,
    TelemetryData,
    AnonymousAuthProvider,
    IAuthProviderCallGuard,
} from "@gooddata/sdk-backend-base";
import { DateFormatter } from "../convertors/fromBackend/dateFormatting/types.js";
import { defaultDateFormatter } from "../convertors/fromBackend/dateFormatting/defaultDateFormatter.js";
import { TigerOrganization, TigerOrganizations } from "./organization/index.js";
import { LIB_VERSION, LIB_NAME } from "../__version.js";
import { TigerSpecificFunctions, buildTigerSpecificFunctions } from "./tigerSpecificFunctions.js";
import { TigerEntitlements } from "./entitlements/index.js";
import { TigerDataSourcesService } from "./dataSources/index.js";

const CAPABILITIES: IBackendCapabilities = {
    hasTypeScopedIdentifiers: true,
    canCalculateGrandTotals: true,
    canCalculateSubTotals: true,
    canCalculateNativeTotals: false,
    canCalculateTotals: true,
    canExportCsv: true,
    canExportXlsx: true,
    canSortData: false,
    canTransformExistingResult: false,
    canWorkspaceManagerSeeEverySharedObject: true,
    maxDimensions: 2,
    supportsElementUris: false,
    supportsObjectUris: false,
    supportsCsvUploader: false,
    supportsRankingFilter: true,
    supportsRankingFilterWithMeasureValueFilter: false,
    supportsElementsQueryParentFiltering: true,
    supportsKpiWidget: false,
    supportsWidgetEntity: false,
    supportsHyperlinkAttributeLabels: true,
    supportsGenericDateAttributeElements: true,
    supportsExplain: true,
    supportsAccessControl: true,
    usesStrictAccessControl: true,
    supportsOwners: true,
    allowsInconsistentRelations: true,
    supportsTimeGranularities: true,
    supportsHierarchicalWorkspaces: true,
    supportsCustomColorPalettes: true,
    supportsOrganizationSettings: true,
    supportsInlineMeasures: true,
    supportsBootstrapResource: false,
    supportsMetadataObjectLocking: false,
    supportsGranularAccessControl: true,
    supportsEveryoneUserGroupForAccessControl: true,
    supportsNonProductionDatasets: false,
    supportsShowAllAttributeValues: true,
    supportsSeparateLatitudeLongitudeLabels: true,
    supportsEnumeratingDatetimeAttributes: false,
    supportsHiddenAndLockedFiltersOnUI: true,
    supportsAttributeHierarchies: true,
    supportsSettingConnectingAttributes: false,
    supportsKeepingDependentFiltersSelection: true,
    supportsCircularDependencyInFilters: true,
    allowMultipleInteractionsPerAttributeAndMeasure: true,
    supportsShowingFilteredElements: true,
    supportsSingleSelectDependentFilters: true,
    supportsCrossFiltering: true,
    supportsMultipleDateFilters: true,
    supportsAttributeFilterElementsLimiting: true,
    supportsRichTextWidgets: true,
};

/**
 * Client-specific configuration for the tiger backend allows to specify additional telemetry information.
 *
 * @public
 */
export type TigerBackendConfig = {
    /**
     * Name of frontend package, this will be recorded by backend as initiator of HTTP requests.
     */
    packageName?: string;

    /**
     * Version of the frontend package, this will be recorded by backend as initiator of HTTP requests.
     */
    packageVersion?: string;

    /**
     * Function used to format date values for a given granularity. It is given a parsed Date value and an appropriate granularity.
     * If not specified, a default date formatted will be used.
     */
    dateFormatter?: DateFormatter;
};

/**
 * Provides a way for the TigerBackend to expose some of its backend specific functions.
 */
type TigerSpecificFunctionsSubscription = {
    onTigerSpecificFunctionsReady?: (functions: TigerSpecificFunctions) => void;
    onContractExpired?: (tier: string) => void;
};

/**
 * An implementation of analytical backend for GoodData CloudNative (codename tiger).
 */
export class TigerBackend implements IAnalyticalBackend {
    public readonly capabilities: IBackendCapabilities = CAPABILITIES;
    public readonly config: IAnalyticalBackendConfig;

    private readonly telemetry: TelemetryData;
    private readonly implConfig: TigerBackendConfig & TigerSpecificFunctionsSubscription;
    private readonly client: ITigerClient;
    private readonly authProvider: IAuthProviderCallGuard;
    private readonly dateFormatter: DateFormatter;

    constructor(
        config: IAnalyticalBackendConfig = {},
        implConfig: TigerBackendConfig & TigerSpecificFunctionsSubscription = {},
        telemetry: TelemetryData = {},
        authProvider?: IAuthProviderCallGuard,
    ) {
        this.config = config;
        this.implConfig = implConfig;
        this.telemetry = telemetry;
        this.authProvider = authProvider || new AnonymousAuthProvider();
        this.dateFormatter = implConfig.dateFormatter ?? defaultDateFormatter;

        const axios = createAxios(this.config, this.implConfig, this.telemetry);
        interceptBackendErrorsToConsole(axios);

        this.client = tigerClientFactory(axios);

        this.authProvider.initializeClient?.(this.client);

        if (this.implConfig.onTigerSpecificFunctionsReady) {
            const specificFunctions = buildTigerSpecificFunctions(this, this.authApiCall);
            this.implConfig.onTigerSpecificFunctionsReady(specificFunctions);
        }
    }

    public onHostname(hostname: string): IAnalyticalBackend {
        return new TigerBackend({ ...this.config, hostname }, this.implConfig, this.telemetry);
    }

    public withTelemetry(componentName: string, props: object): IAnalyticalBackend {
        return new TigerBackend(
            this.config,
            this.implConfig,
            { componentName, props: Object.keys(props) },
            this.authProvider,
        );
    }

    public withAuthentication(provider: IAuthenticationProvider): IAnalyticalBackend {
        const guardedAuthProvider = new AuthProviderCallGuard(provider);

        return new TigerBackend(this.config, this.implConfig, this.telemetry, guardedAuthProvider);
    }

    public deauthenticate(): Promise<void> {
        if (!this.authProvider) {
            throw new NotAuthenticated("Backend is not set up with authentication provider.");
        }
        return this.authProvider.deauthenticate(this.getAuthenticationContext());
    }

    public organization(organizationId: string): IOrganization {
        return new TigerOrganization(this.authApiCall, organizationId);
    }

    public organizations(): IOrganizations {
        return new TigerOrganizations(this.authApiCall);
    }

    public entitlements(): IEntitlements {
        return new TigerEntitlements(this.authApiCall);
    }

    public dataSources(): IDataSourcesService {
        return new TigerDataSourcesService(this.authApiCall);
    }

    public currentUser(): IUserService {
        return new TigerUserService(this.authApiCall);
    }

    public workspace(id: string): IAnalyticalWorkspace {
        invariant(isString(id), `Invalid workspaceId, expected a string, got: ${id}`);
        return new TigerWorkspace(this.authApiCall, id, this.dateFormatter);
    }

    public workspaces(): IWorkspacesQueryFactory {
        return new TigerWorkspaceQueryFactory(this.authApiCall, this.dateFormatter);
    }

    public isAuthenticated = async (): Promise<IAuthenticatedPrincipal | null> => {
        try {
            // the return await is crucial here so that we also catch the async errors
            return await this.authProvider.getCurrentPrincipal({ client: this.client, backend: this });
        } catch (err) {
            if (isNotAuthenticatedResponse(err) || isNotAuthenticated(err)) {
                return null;
            }
            throw err;
        }
    };

    public authenticate = async (force: boolean): Promise<IAuthenticatedPrincipal> => {
        if (!force) {
            return this.authApiCall(async (client) => {
                const principal = await this.authProvider.getCurrentPrincipal({ client, backend: this });
                invariant(principal, "Principal must be defined");
                return principal;
            });
        }

        try {
            // the return await is crucial here so that we also catch the async errors
            return await this.triggerAuthentication(true);
        } catch (err) {
            invariant(isError(err)); // if this bombs, the code in the try block threw something strange
            throw this.handleAnalyticalBackendError(convertApiError(err));
        }
    };

    /**
     * Perform API call that requires authentication. The call will be decorated with error handling
     * such that not authenticated errors will trigger authentication flow AND other errors will be
     * converted using the provided converter and throw.
     *
     * @param call - a call which requires an authenticated session
     * @param errorConverter - converter from rest client errors to analytical backend errors
     */
    private authApiCall = async <T>(
        call: AuthenticatedAsyncCall<ITigerClient, T>,
        errorConverter: ErrorConverter = convertApiError,
    ): Promise<T> => {
        try {
            // the return await is crucial here so that we also catch the async errors
            return await call(this.client, await this.getAsyncCallContext());
        } catch (err) {
            invariant(isError(err)); // if this bombs, the code in the try block threw something strange

            // if we receive some other error than missing auth, we fail fast: no need to try the auth
            // one more time, since it was not the problem in the first place
            if (!isNotAuthenticatedResponse(err)) {
                throw this.handleAnalyticalBackendError(errorConverter(err));
            }

            // else we try to trigger the authentication once more and then we repeat the original call
            // with the newly obtained async call context
            try {
                await this.triggerAuthentication();
                // the return await is crucial here so that we also catch the async errors
                return await call(this.client, await this.getAsyncCallContext());
            } catch (err2) {
                invariant(isError(err2)); // if this bombs, the code in the try block threw something strange
                throw this.handleAnalyticalBackendError(errorConverter(err2));
            }
        }
    };

    /**
     * Triggers relevant handler if the provided error is an instance of
     * {@link @gooddata/sdk-backend-spi#NotAuthenticated} or {@link @gooddata/sdk-backend-spi#ContractExpired}.
     *
     * @param err - error to observe and trigger handler for
     * @returns the original error to facilitate re-throwing
     */
    private handleAnalyticalBackendError = <T>(err: T): T => {
        if (isNotAuthenticated(err)) {
            this.authProvider.onNotAuthenticated?.({ client: this.client, backend: this }, err);
        } else if (isContractExpired(err)) {
            this.implConfig.onContractExpired?.(err.message);
        }

        return err;
    };

    private getAuthenticationContext = (): IAuthenticationContext => {
        return { client: this.client, backend: this };
    };

    private getAsyncCallContext = async (): Promise<IAuthenticatedAsyncCallContext> => {
        const getPrincipal = async (): Promise<IAuthenticatedPrincipal> => {
            if (!this.authProvider) {
                throw new NotAuthenticated("Cannot obtain principal without an authProvider.");
            }

            const principal = await this.authProvider.getCurrentPrincipal({
                client: this.client,
                backend: this,
            });

            return principal ?? this.authProvider.authenticate(this.getAuthenticationContext());
        };

        return {
            getPrincipal,
        };
    };

    private triggerAuthentication = (reset = false): Promise<IAuthenticatedPrincipal> => {
        if (!this.authProvider) {
            return Promise.reject(
                new NotAuthenticated("Backend is not set up with authentication provider."),
            );
        }

        if (reset) {
            this.authProvider.reset();
        }

        return this.authProvider.authenticate({ client: this.client, backend: this });
    };
}

function createAxios(
    config: IAnalyticalBackendConfig,
    implConfig: TigerBackendConfig,
    telemetry: TelemetryData,
): AxiosInstance {
    const baseUrl = config.hostname ? config.hostname : undefined;
    const headers = createHeaders(implConfig, telemetry);

    return newAxios(baseUrl, headers);
}

function interceptBackendErrorsToConsole(client: AxiosInstance): AxiosInstance {
    client.interceptors.response.use(identity, (error) => {
        const response: AxiosResponse = error.response;

        // If there is no response object (for example for blocked requests), print the whole error.
        if (!response) {
            console.error("Tiger backend threw an error:", error);
        }
        // Else if the response is an object (JSON parsed by axios) and there is a problem, then log error
        // into console for easier diagnostics.
        else if (inRange(response.status, 400, 600) && typeof response.data === "object") {
            // Title is redundant (Bad Request)
            const details = omit(response.data, ["title"]);
            console.error("Tiger backend threw an error:", details);
        }

        return Promise.reject(error);
    });

    return client;
}

function createHeaders(implConfig: TigerBackendConfig, telemetry: TelemetryData): { [name: string]: string } {
    const headers: { [name: string]: string } = {
        "X-GDC-JS-PACKAGE": LIB_NAME,
        "X-GDC-JS-PACKAGE-VERSION": LIB_VERSION,
    };

    if (telemetry.componentName) {
        headers["X-GDC-JS-SDK-COMP"] = telemetry.componentName;
    }

    if (telemetry.props && !isEmpty(telemetry.props)) {
        headers["X-GDC-JS-SDK-COMP-PROPS"] = telemetry.props.join(",");
    }

    if (implConfig.packageName && implConfig.packageVersion) {
        headers["X-GDC-JS-PACKAGE"] = implConfig.packageName;
        headers["X-GDC-JS-PACKAGE-VERSION"] = implConfig.packageVersion;
    }

    return headers;
}

function isNotAuthenticatedResponse(err: any): boolean {
    return err?.response?.status === 401;
}
