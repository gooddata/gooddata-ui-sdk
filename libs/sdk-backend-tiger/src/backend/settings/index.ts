// (C) 2020-2023 GoodData Corporation
import { TigerSettingsService } from "./settings.js";
import { mapTypeToKey } from "./mapping.js";

export { TigerSettingsService, mapTypeToKey };
