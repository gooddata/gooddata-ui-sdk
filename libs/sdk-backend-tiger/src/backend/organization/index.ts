// (C) 2021-2024 GoodData Corporation

import {
    IOrganization,
    IOrganizations,
    IOrganizationSettingsService,
    IOrganizationStylingService,
    ISecuritySettingsService,
    IOrganizationUserService,
    IOrganizationPermissionService,
    IOrganizationNotificationChannelService,
    IOrganizationLlmEndpointsService,
    IOrganizationNotificationService,
} from "@gooddata/sdk-backend-spi";
import { IOrganizationDescriptor, idRef, IOrganizationDescriptorUpdate } from "@gooddata/sdk-model";

import { SecuritySettingsService } from "./securitySettings.js";
import { TigerAuthenticatedCallGuard } from "../../types/index.js";
import { OrganizationStylingService } from "./styling.js";
import { OrganizationSettingsService } from "./settings.js";
import { OrganizationUsersService } from "./users.js";
import { OrganizationPermissionService } from "./permissions.js";
import { OrganizationNotificationChannelService } from "./notificationChannels.js";
import { OrganizationLlmEndpointsService } from "./llmEndpoints.js";
import { OrganizationNotificationService } from "./notifications.js";

export class TigerOrganization implements IOrganization {
    constructor(
        private readonly authCall: TigerAuthenticatedCallGuard,
        public readonly organizationId: string,
        public readonly organizationName?: string,
    ) {}

    public async getDescriptor(includeAdditionalDetails?: boolean): Promise<IOrganizationDescriptor> {
        // if we already have the data, no reason to download them again
        if (this.organizationName && !includeAdditionalDetails) {
            return {
                id: this.organizationId,
                title: this.organizationName,
            };
        }

        if (includeAdditionalDetails) {
            const result = await this.authCall((client) =>
                client.entities.getEntityOrganizations({
                    id: this.organizationId,
                    // we are interested only in these for now (can be extended in future)
                    include: ["bootstrapUser", "bootstrapUserGroup"],
                }),
            );

            // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
            const organizationName = result.data.data.attributes?.name!;
            // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
            const bootstrapUser = result.data.data?.relationships?.bootstrapUser?.data!;
            // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
            const bootstrapUserGroup = result.data.data?.relationships?.bootstrapUserGroup?.data!;

            return {
                id: this.organizationId,
                title: organizationName,
                bootstrapUser: idRef(bootstrapUser.id, bootstrapUser.type),
                bootstrapUserGroup: idRef(bootstrapUserGroup.id, bootstrapUserGroup.type),
                earlyAccess: result.data.data.attributes?.earlyAccess ?? undefined,
                earlyAccessValues: result.data.data.attributes?.earlyAccessValues ?? undefined,
            };
        }

        const { organizationName, organizationId } = await this.authCall((client) =>
            client.profile.getCurrent(),
        );
        return {
            id: organizationId,
            title: organizationName,
        };
    }

    public async updateDescriptor(
        descriptor: IOrganizationDescriptorUpdate,
    ): Promise<IOrganizationDescriptor> {
        const result = await this.authCall((client) =>
            client.entities.patchEntityOrganizations({
                id: this.organizationId,
                include: ["bootstrapUser", "bootstrapUserGroup"],
                jsonApiOrganizationPatchDocument: {
                    data: {
                        id: this.organizationId,
                        type: "organization",
                        attributes: {
                            name: descriptor.title,
                            // type casts are necessary because nulls are not allowed in the type definition,
                            // but backend expects them in case we want to delete the value
                            earlyAccess: descriptor.earlyAccess as string | undefined,
                            earlyAccessValues: descriptor.earlyAccessValues as string[] | undefined,
                        },
                    },
                },
            }),
        );

        const resultData = result.data.data;
        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
        const organizationName = resultData.attributes?.name!;
        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
        const bootstrapUser = resultData.relationships?.bootstrapUser?.data!;
        // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
        const bootstrapUserGroup = resultData.relationships?.bootstrapUserGroup?.data!;
        const earlyAccess = resultData.attributes?.earlyAccess ?? undefined;
        const earlyAccessValues = resultData.attributes?.earlyAccessValues ?? undefined;

        return {
            id: this.organizationId,
            title: organizationName,
            bootstrapUser: idRef(bootstrapUser.id, bootstrapUser.type),
            bootstrapUserGroup: idRef(bootstrapUserGroup.id, bootstrapUserGroup.type),
            earlyAccess,
            earlyAccessValues,
        };
    }

    public securitySettings(): ISecuritySettingsService {
        return new SecuritySettingsService(this.organizationId);
    }

    public styling(): IOrganizationStylingService {
        return new OrganizationStylingService(this.authCall);
    }

    public settings(): IOrganizationSettingsService {
        return new OrganizationSettingsService(this.authCall);
    }

    public users(): IOrganizationUserService {
        return new OrganizationUsersService(this.authCall);
    }

    public permissions(): IOrganizationPermissionService {
        return new OrganizationPermissionService(this.authCall);
    }

    public notificationChannels(): IOrganizationNotificationChannelService {
        return new OrganizationNotificationChannelService(this.authCall);
    }

    public llmEndpoints(): IOrganizationLlmEndpointsService {
        return new OrganizationLlmEndpointsService(this.authCall);
    }

    public notifications(): IOrganizationNotificationService {
        return new OrganizationNotificationService(this.authCall);
    }
}

export class TigerOrganizations implements IOrganizations {
    constructor(private readonly authCall: TigerAuthenticatedCallGuard) {}

    public async getCurrentOrganization(): Promise<IOrganization> {
        const { organizationName, organizationId } = await this.authCall((client) =>
            client.profile.getCurrent(),
        );
        return new TigerOrganization(this.authCall, organizationId, organizationName);
    }
}
