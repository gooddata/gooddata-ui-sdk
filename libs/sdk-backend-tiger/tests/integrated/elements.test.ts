// (C) 2022-2024 GoodData Corporation
import { testBackend, testWorkspace } from "./backend.js";
import {
    attributeDisplayFormRef,
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    IAttributeElement,
} from "@gooddata/sdk-model";
import omit from "lodash/omit.js";

import * as Md from "../../src/fixtures/full.js";
import * as ReferenceRecords from "../../src/fixtures/referenceObjects.js";
import { describe, expect, it, beforeAll } from "vitest";

const backend = testBackend();

beforeAll(async () => {
    await backend.authenticate(true);
});

describe("tiger elements", () => {
    describe("forDisplayForm", () => {
        it("should load attribute elements for existing display form", async () => {
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forDisplayForm(attributeDisplayFormRef(Md.AccountId))
                .withLimit(20)
                .query();

            expect(omit(result, "cacheId")).toMatchSnapshot();
        });
    });

    describe("forFilter", () => {
        const testAttributeElement = (ReferenceRecords.productName as IAttributeElement[]).find(
            (el) => el.title === "Educationly",
        );

        const testAttributeElementUri = testAttributeElement!.uri!.replace(
            "referenceworkspace",
            testWorkspace(),
        );

        it("should load attribute filter elements for provided positive attribute filter", async () => {
            const attributeFilter = newPositiveAttributeFilter(Md.Product.Name, {
                uris: [testAttributeElementUri],
            });
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forFilter(attributeFilter)
                .query();

            expect(result).toMatchSnapshot();
        });

        it("should load attribute filter elements for provided negative attribute filter", async () => {
            const testAttributeRef = attributeDisplayFormRef(Md.Product.Name);
            const attributeFilter = newNegativeAttributeFilter(testAttributeRef, {
                uris: [testAttributeElementUri],
            });
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forFilter(attributeFilter)
                .query();

            expect(result).toMatchSnapshot();
        });

        it("should load no attribute filter elements for provided ALL attribute filter", async () => {
            const testAttributeRef = attributeDisplayFormRef(Md.Product.Name);
            const allAttributeFilter = newNegativeAttributeFilter(testAttributeRef, {
                uris: [],
            });
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forFilter(allAttributeFilter)
                .query();

            expect(result).toMatchSnapshot();
        });

        it("should return attribute filter elements for provided attribute filter with elements by value", async () => {
            const testAttributeRef = attributeDisplayFormRef(Md.Product.Name);
            const attributeFilter = newNegativeAttributeFilter(testAttributeRef, {
                values: ["Educationly"],
            });
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forFilter(attributeFilter)
                .query();

            expect(result).toMatchSnapshot();
        });

        it("should return correct page for given displayForm", async () => {
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forDisplayForm(attributeDisplayFormRef(Md.Account.Default))
                .withLimit(2)
                .query();
            const page = await result.goTo(3);
            expect(omit(page, "items", "cacheId")).toMatchSnapshot();
        });

        it("should return empty result for out-of-range page in initial request", async () => {
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forDisplayForm(attributeDisplayFormRef(Md.Account.Default))
                .withLimit(100)
                .withOffset(5000)
                .query();

            expect(omit(result, "cacheId")).toMatchSnapshot();
        });

        it("should return empty result for out-of-range page with goTo", async () => {
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forDisplayForm(attributeDisplayFormRef(Md.Account.Default))
                .withLimit(100)
                .query();

            const page = await result.goTo(100);
            expect(omit(page, "cacheId")).toMatchSnapshot();
        });

        it("should return empty result for out-of-range page with next", async () => {
            const result = await backend
                .workspace(testWorkspace())
                .attributes()
                .elements()
                .forDisplayForm(attributeDisplayFormRef(Md.Account.Default))
                .withLimit(100)
                .withOffset(4840)
                .query();

            const page = await result.next();
            expect(omit(page, "cacheId")).toMatchSnapshot();
        });
    });
});
