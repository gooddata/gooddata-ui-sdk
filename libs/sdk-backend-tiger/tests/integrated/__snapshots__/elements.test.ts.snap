// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tiger elements > forDisplayForm > should load attribute elements for existing display form 1`] = `
{
  "all": [Function],
  "allSorted": [Function],
  "getData": [Function],
  "goTo": [Function],
  "items": [
    {
      "title": "38",
      "uri": "38",
    },
    {
      "title": "78",
      "uri": "78",
    },
    {
      "title": "224",
      "uri": "224",
    },
    {
      "title": "545",
      "uri": "545",
    },
    {
      "title": "552",
      "uri": "552",
    },
    {
      "title": "955129",
      "uri": "955129",
    },
    {
      "title": "955131",
      "uri": "955131",
    },
    {
      "title": "955133",
      "uri": "955133",
    },
    {
      "title": "955139",
      "uri": "955139",
    },
    {
      "title": "955141",
      "uri": "955141",
    },
    {
      "title": "955143",
      "uri": "955143",
    },
    {
      "title": "955145",
      "uri": "955145",
    },
    {
      "title": "955147",
      "uri": "955147",
    },
    {
      "title": "955153",
      "uri": "955153",
    },
    {
      "title": "955155",
      "uri": "955155",
    },
    {
      "title": "955157",
      "uri": "955157",
    },
    {
      "title": "955159",
      "uri": "955159",
    },
    {
      "title": "955163",
      "uri": "955163",
    },
    {
      "title": "955165",
      "uri": "955165",
    },
    {
      "title": "955167",
      "uri": "955167",
    },
  ],
  "limit": 20,
  "next": [Function],
  "offset": 0,
  "totalCount": 4846,
}
`;

exports[`tiger elements > forFilter > should load attribute filter elements for provided negative attribute filter 1`] = `
InMemoryPaging {
  "allItems": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "items": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "limit": 100,
  "offset": 0,
  "totalCount": 1,
}
`;

exports[`tiger elements > forFilter > should load attribute filter elements for provided positive attribute filter 1`] = `
InMemoryPaging {
  "allItems": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "items": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "limit": 100,
  "offset": 0,
  "totalCount": 1,
}
`;

exports[`tiger elements > forFilter > should load no attribute filter elements for provided ALL attribute filter 1`] = `
InMemoryPaging {
  "allItems": [],
  "items": [],
  "limit": 100,
  "offset": 0,
  "totalCount": 0,
}
`;

exports[`tiger elements > forFilter > should return attribute filter elements for provided attribute filter with elements by value 1`] = `
InMemoryPaging {
  "allItems": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "items": [
    {
      "title": "Educationly",
      "uri": "Educationly",
    },
  ],
  "limit": 100,
  "offset": 0,
  "totalCount": 1,
}
`;

exports[`tiger elements > forFilter > should return correct page for given displayForm 1`] = `
{
  "all": [Function],
  "allSorted": [Function],
  "getData": [Function],
  "goTo": [Function],
  "limit": 2,
  "next": [Function],
  "offset": 6,
  "totalCount": 4846,
}
`;

exports[`tiger elements > forFilter > should return empty result for out-of-range page in initial request 1`] = `
{
  "all": [Function],
  "allSorted": [Function],
  "getData": [Function],
  "goTo": [Function],
  "items": [],
  "limit": 100,
  "next": [Function],
  "offset": 5000,
  "totalCount": 4846,
}
`;

exports[`tiger elements > forFilter > should return empty result for out-of-range page with goTo 1`] = `
{
  "all": [Function],
  "allSorted": [Function],
  "getData": [Function],
  "goTo": [Function],
  "items": [],
  "limit": 100,
  "next": [Function],
  "offset": 10000,
  "totalCount": 4846,
}
`;

exports[`tiger elements > forFilter > should return empty result for out-of-range page with next 1`] = `
{
  "all": [Function],
  "allSorted": [Function],
  "getData": [Function],
  "goTo": [Function],
  "items": [],
  "limit": 100,
  "next": [Function],
  "offset": 4940,
  "totalCount": 4846,
}
`;
